### PR TITLE
Promotion and String Updates/Fixes

### PR DESCRIPTION
**Pull request by jdconrad**
* Improved type promotion for def type.
* Fix for numeric overflow bug in compound assignment.
* Fix for string concatenation.
* Added significant number of tests for def type related to numerical ops.
* Allow string constants to be used with method such as "x".equals("x")
* Added new tests for the string constants.
* Separate variable from field at the grammar level to give them separate nodes in the tree for processing.

### CODE REVIEW

File: src/main/java/org/elasticsearch/plan/a/Utility.java [hunk 1]
```diff
@@ -511,7 +531,130 @@ public static short toShortExact(int value) {
...
+        if (s != value) {
+            throw new ArithmeticException("short overflow");
+        }
+        return s;
+    }
+
+    /**
+     * Like {@link Math#toIntExact(long)} but for short range.
+     */
+    public static short toShortExact(float value) {
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/plan/a/Utility.java]: its no longer really exact though right? maybe the float/double variants should be named "withoutOverflow"? Also we should keep in mind both checks will fail for a NaN input. I don't think its important to have a NaN check, unless the previous operation could have produced NaN from two finite values.

**Comment by rmuir** [comment]
looks good. I added one note. sorry for delays in reviewing.

File: src/main/java/org/elasticsearch/plan/a/Utility.java (hunk 1 above)
Comment (jdconrad) [inline src/main/java/org/elasticsearch/plan/a/Utility.java]: That's a good point.  I will rename these.

**Comment by jdconrad** [comment]
Thanks for the review!

**Pull request merged**